Cross compilation fixes [april 2020]

matthewbauer opened the pull request:
More cross-compilation fixes. 

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

matthewbauer committed:
gdk-pixbuf: support cross compilation

- add glib to nativeBuildInputs
- only update loaders cache on native builds

matthewbauer committed:
epiphany: include gtk3 build dependency

matthewbauer committed:
glib: put gsettings directory in target

These should be runtime, not build time deps

matthewbauer committed:
hicolor-icon-theme: only add target icons to deps

matthewbauer committed:
treewide/nixos: use stdenv.cc.libc instead of glibc when available

This prevents duplication in cross-compiled nixos machines. The
bootstrapped glibc differs from the natively compiled one, so we get
two glibc’s in the closure. To reduce closure size, just use
stdenv.cc.libc where available.

matthewbauer committed:
gdk-pixbuf: make target env hook

Unfortunately, we need to emulate the system to get a real cache.
Native version doesn’t know the right paths.

matthewbauer committed:
plymouth: set systemd-tty-ask-password-agent path

This is needed in cross where systemd is not in path.

matthewbauer committed:
libXres: enable cross-compilation

- adds utilmacros
- set null malloc flag

matthewbauer committed:
fontconfig: only generate cache on native compilation

We can’t cross-compile the cache, so just skip it for now.

matthewbauer committed:
libvpx: set proper target name

- Needs to be like <cpu>-<os>-gcc.
- Remove unneeded --enable-external-build flag
    This confuses the feature detection and is not needed for cross
    compilation anymore.

matthewbauer committed:
xorg.libXScrnSaver: add utilmacros and malloc 0

matthewbauer committed:
libva: add wayland as native build input

Also set wayland-scanner to proper path

matthewbauer committed:
bluez: don’t build python packages when tests are disabled

Can’t run these on cross anyway

matthewbauer committed:
alsa-plugins: move pkgconfig to native

matthewbauer committed:
libvisual: fix null malloc check on cross

this is needed to build

matthewbauer committed:
libasyncns: override malloc null check in cross

This is needed to build and avoid malloc error.

matthewbauer committed:
cracklib: only use local binaries on native build

matthewbauer committed:
retroarch: add wayland dependency and fix cross

- Need custom pkg-config setting
- add wayland & libxkbcommon

matthewbauer committed:
vulkan-loader: make cmake a native build input

matthewbauer committed:
git: fix cross compilation

matthewbauer committed:
waylandpp: fix cross compilation

- add libffi dependency
- provide wayland-scanner++ when cross compiling

matthewbauer committed:
tinyxml2: fix cross compilation

- use native unzip
- add target prefix to LD

matthewbauer committed:
libcrossguid: get prefixed utility

matthewbauer committed:
setup-hooks/auto-patchelf.sh: get prefixed readelf

matthewbauer committed:
libcec: get native cmake for cross

matthewbauer committed:
libcdio: get native help2man

matthewbauer committed:
libcddb: force malloc and realloc

This needs to be set to avoid having undefined malloc symbol.

matthewbauer committed:
ftgl: add --with-ft-prefix for finding freetype

freetype-config is not in PATH when cross-compiling. We need to
provide our own.

matthewbauer committed:
flatbuffers: don’t build tests when cross compiling

matthewbauer committed:
fstrcmp: patch for cross

Patches fstrcmp to cross-compile

matthewbauer committed:
kodi: fix cross compilation

- provide native binaries
- set cpu / os / etc. flags
- fix wayland compilation

doronbehar commented:
@matthewbauer I just came about https://github.com/NixOS/nixpkgs/issues/83575 which suggests that `dvtm` should have `ncurses` in it's native build inputs (or alike). Is there a chance you could take a look at that and perhaps fix it in this PR as well? Thanks.

matthewbauer committed:
dvtm: add ncurses dev input to PATH

fixes cross compilation

matthewbauer committed:
python-dbusmock: avoid eval error when missing test output of bluez

This fixes an evaluation error that occurs when we don’t have
bluez.test. getOutput defaults to "out" in these cases.

Mic92 commented on pkgs/development/libraries/libvpx/default.nix:
```diff
@@ -135,16 +135,10 @@ stdenv.mkDerivation rec {
     # See all_platforms: https://github.com/webmproject/libvpx/blob/master/configure
     # Darwin versions: 10.4=8, 10.5=9, 10.6=10, 10.7=11, 10.8=12, 10.9=13, 10.10=14
-    "--force-target=${stdenv.hostPlatform.config}${
+    "--force-target=${stdenv.hostPlatform.parsed.cpu.name}-${stdenv.hostPlatform.parsed.kernel.name}${
```
Nice. I strugged with this one. I think so on aarch64 it is still incorrect because `arm64` != `aarch64`.

matthewbauer replied:
I was able to build for aarch64. I think they rewrite it here:

https://github.com/webmproject/libvpx/blob/0c46742f56d06bf284c781bd754eff3b1d97c3c8/build/make/configure.sh#L723

Mic92 commented on pkgs/applications/version-management/git-and-tools/git/default.nix:
```diff
@@ -88,7 +90,7 @@ stdenv.mkDerivation {
   makeFlags = [
     "prefix=\${out}"
-    "SHELL_PATH=${stdenv.shell}"
+    "SHELL_PATH=${if (stdenv.hostPlatform != stdenv.buildPlatform) then "/bin/sh" else stdenv.shell}"
```
```suggestion
    "SHELL_PATH=${runtimeShell}"
```

Mic92 replied:
Just read the commit message... The only pure alternative to your approach would be to replace `stdenv.shell` with `runtimeShell` in `$out`.

Mic92 replied:
Otherwise if we have a sandbox build `/bin/sh` is reasonable reproducible, could add the reasoning as a comment above for a future reader?

Profpatsch replied:
We are referencing `"/bin/sh"` in nixpkgs … oh no.

Ericson2314 replied:
Yeah if it is about the old vs new version of the code (stateful) I like to put in the commit message, but if it is about the new version of the code alone, I rather it be a comment.

Comments have a prayer of being edited to stay up to date :), but commit messages only describe a moment in time.

matthewbauer replied:
Yeah - patch-shebangs.sh bails out when the path is in the nix store.

I'll leave this change out of this set for now. Using /bin/sh is definitely not good, so some other fix is needed. Git will still build, you just can't run some of its shell scripts.

wmertens replied:
:thinking: isn't this a general problem with cross builds, for e.g. Python install scripts?

How about using a substitution set for buildDeps that should be replaced with runDeps? And maybe automatically testing for buildDeps that are referenced in the output but not specifically allowed?

Mic92 commented on nixos/modules/config/fonts/fontconfig.nix:
```diff
@@ -60,11 +60,13 @@ let
       <fontconfig>
         <!-- Font directories -->
         ${concatStringsSep "\n" (map (font: "<dir>${font}</dir>") config.fonts.fonts)}
+        ${optionalString (pkgs.stdenv.hostPlatform == pkgs.stdenv.buildPlatform) ''
```
Also please add a comment here.

Ericson2314 commented on pkgs/development/libraries/librsvg/default.nix:
```diff
@@ -20,14 +24,26 @@ stdenv.mkDerivation rec {
+    [target."${toRustTarget stdenv.hostPlatform}"]
+    "linker" = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}cc"
+    EOF
+  '';
```
Can we share this with the other places we do this?

Mic92 replied:
Some setup hook would be nice if it is not possible to hard code this during compilation.
